File: 0011.  Container With Most Water/0011.java
// time complexity: O(n)
// space complexity: O(1)
class Solution {
    public int maxArea(int[] height) {
        // height contains at least 2 elements.
        int max_area = 0, current_max = 0;
        int i = 0, j = height.length-1;
        while(i!=j){
            // find the current max
            current_max = Math.min(height[j], height[i]) * (j-i);
            // and compare if it's greater than max area
            max_area = Math.max(max_area, current_max);

            // move the pointer to the shorter height
            if (height[i] > height[j])
                j--;
            else i++;
        }

        return max_area;
    }
}

File: 0011.  Container With Most Water/0011.py
# time complexity: O(n)
# space complexity: O(1)
class Solution(object):
    def maxArea(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        left = 0
        right = len(height) - 1

        max_area = 0
        current_area = 0

        while left != right:
            current_area = min(height[left], height[right]) * (right-left)
            max_area = max(max_area, current_area)

            # move the shorter pointer
            if height[left] < height[right]:
                left+=1
            else:
                right-=1
        
        return max_area